finstmt/bs/data.py

```diff
@@ -1,53 +1,98 @@
-    deposit_liab: Optional[float] = 0
-    other_lt_liab: Optional[float] = 0
-    common_stock: Optional[float] = 0
-    minority_interest: Optional[float] = 0
-    other_income: Optional[float] = 0
-    retained_earnings: Optional[float] = 0
-
-    cash_and_st_invest: Optional[float] = None
-    total_current_assets: Optional[float] = None
-    net_ppe: Optional[float] = None
-    total_non_current_assets: Optional[float] = None
-    total_assets: Optional[float] = None
-    total_current_liab: Optional[float] = None
-    total_debt: Optional[float] = None
-    total_non_current_liab: Optional[float] = None
-    total_liab: Optional[float] = None
-    total_equity: Optional[float] = None
-    total_liab_and_equity: Optional[float] = None
-
-    items_config = BALANCE_SHEET_INPUT_ITEMS
-
-    @property
-    def nwc(self):
-        return self.receivables + self.inventory - self.payables
+    t = symbols("t", cls=Idx)
+
+    def __init__(self, *args, **kwargs):
+        _fields = [
+            (
+                item.key,
+                numpy.float64,
+                field(default=0, repr=(False if item.key == "nwc" else True)),
+            )
+            for item in self.items_config_list
+        ]
+        sort_order = {
+            "cash": 0,
+            "st_invest": 1,
+            "receivables": 2,
+            "inventory": 3,
+            "lt_invest": 4,
+            "def_tax_st": 5,
+            "other_current_assets": 6,
+            "gross_ppe": 7,
+            "dep": 8,
+            "goodwill": 9,
+            "def_tax_lt": 10,
+            "other_lt_assets": 11,
+            "payables": 12,
+            "current_lt_debt": 13,
+            "st_debt": 14,
+            "lt_debt": 15,
+            "deferred_rev": 16,
+            "tax_liab_st": 17,
+            "other_current_liab": 18,
+            "tax_liab_lt": 19,
+            "deposit_liab": 20,
+            "other_lt_liab": 21,
+            "common_stock": 22,
+            "minority_interest": 23,
+            "other_income": 24,
+            "retained_earnings": 25,
+            "cash_and_st_invest": 26,
+            "total_current_assets": 27,
+            "net_ppe": 28,
+            "total_non_current_assets": 29,
+            "total_assets": 30,
+            "total_current_liab": 31,
+            "total_debt": 32,
+            "total_non_current_liab": 33,
+            "total_liab": 34,
+            "total_equity": 35,
+            "total_liab_and_equity": 36,
+            "nwc": 37,
+        }
+        _fields.sort(key=lambda tup: sort_order[tup[0]])
+        MyClass = make_dataclass(
+            "BalanceSheetData",
+            fields=_fields,
+            bases=(FinDataBase,),
+            # nwc now in the config file
+            # namespace={
+            #     "nwc": lambda self: self.receivables + self.inventory - self.payables
+            # },
+        )
+        MyClass.__module__ = "finstmt.bs.data"
+        self.__class__ = MyClass
```

Comment:
Dynamically updating the class instance during instance initialization... interesting idea, I'm not sure if it will have side effects when we construct multiple statements with different configs. 

If we are going the dynamic dataclass route, it seems like the dataclass should be generated in the parent (statement) class and then used, rather than being declared statically separately and then overwritten dynamically during object initialization. 

Or if you can confirm that with this approach setting `self.__class__` does not affect any other instances of the class, then this is fine but we should add comments to explain.

What exactly tied us to using a dataclass? I'm not opposed to some completely different structure as well.

Reply:
class deleted and moved logic to `FinDataBase`